src/core/server/saved_objects/service/lib/legacy_url_aliases/delete_legacy_url_aliases.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as esKuery from '@kbn/es-query';
+
+import type { ISavedObjectTypeRegistry } from '../../../saved_objects_type_registry';
+import type { IndexMapping } from '../../../mappings';
+import { LEGACY_URL_ALIAS_TYPE } from '../../../object_types';
+import type { RepositoryEsClient } from '../repository_es_client';
+import { getSearchDsl } from '../search_dsl';
+import { ALL_NAMESPACES_STRING, DEFAULT_NAMESPACE_STRING } from '../utils';
+
+/** @internal */
+export interface DeleteLegacyUrlAliasesParams {
+  mappings: IndexMapping;
+  registry: ISavedObjectTypeRegistry;
+  client: RepositoryEsClient;
+  getIndexForType: (type: string) => string;
+  /** The object type. */
+  type: string;
+  /** The object ID. */
+  id: string;
+  /**
+   * The namespaces to include or exclude when searching for legacy URL alias targets (depends on the `deleteBehavior` parameter).
+   * Note that using `namespaces: [], deleteBehavior: 'exclusive'` will delete all aliases for this object in all spaces.
+   */
+  namespaces: string[];
+  /**
+   * If this is equal to 'inclusive', all aliases with a `targetNamespace` in the `namespaces` array will be deleted.
+   * If this is equal to 'exclusive', all aliases with a `targetNamespace` _not_ in the `namespaces` array will be deleted.
+   */
+  deleteBehavior: 'inclusive' | 'exclusive';
+}
+
+/**
+ * Deletes legacy URL aliases that point to a given object.
+ *
+ * Note that aliases are only created when an object is converted to become share-capable, and each targetId is deterministically generated
+ * with uuidv5 -- this means that the chances of there actually being _multiple_ legacy URL aliases that target a given type/ID are slim to
+ * none. However, we don't always know exactly what space an alias could be in (if an object exists in multiple spaces, or in all spaces),
+ * so the most straightforward way for us to ensure that aliases are reliably deleted is to use updateByQuery, which is what this function
+ * does.
+ *
+ * @internal
+ */
+export async function deleteLegacyUrlAliases(params: DeleteLegacyUrlAliasesParams) {
+  const { mappings, registry, client, getIndexForType, type, id, namespaces, deleteBehavior } =
+    params;
+
+  if (namespaces.includes(ALL_NAMESPACES_STRING)) {
+    throwError(type, id, '"namespaces" cannot include the * string');
+  }
+
+  // Legacy URL aliases cannot exist in the default space; filter that out
+  const filteredNamespaces = namespaces.filter(
+    (namespace) => namespace !== DEFAULT_NAMESPACE_STRING
+  );
```

Comment:
I wasn't even aware of that tbh...

Reply:
Yeah, aliases are only created when objects in non-default spaces are converted, and I also designed `resolve` with that assumption in mind. Technically we could have chosen to support aliases in the default space (with a `default:` prefix in the raw ID) but there wasn't any practical reason to do so.